File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/view_model/BillEditViewModel.java
package io.github.skywalkerdarren.simpleaccounting.view_model;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import org.joda.time.DateTime;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

import io.github.skywalkerdarren.simpleaccounting.R;
import io.github.skywalkerdarren.simpleaccounting.model.AppRepository;
import io.github.skywalkerdarren.simpleaccounting.model.entity.Account;
import io.github.skywalkerdarren.simpleaccounting.model.entity.Bill;
import io.github.skywalkerdarren.simpleaccounting.model.entity.Type;

/**
 * @author darren
 * @date 2018/4/4
 */

public class BillEditViewModel extends ViewModel {
    private final AppRepository mRepository;
    private final MutableLiveData<String> typeName = new MutableLiveData<>();
    private final MutableLiveData<String> typeImg = new MutableLiveData<>();
    private final MutableLiveData<String> accountImg = new MutableLiveData<>();
    private final MutableLiveData<Integer> accountColor = new MutableLiveData<>(R.color.black);
    private final MutableLiveData<String> balance = new MutableLiveData<>();
    private final MutableLiveData<DateTime> date = new MutableLiveData<>();
    private final MutableLiveData<String> remark = new MutableLiveData<>();
    private final MutableLiveData<UUID> billId = new MutableLiveData<>();
    private final MutableLiveData<UUID> accountId = new MutableLiveData<>();
    private final MutableLiveData<UUID> typeId = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isExpense = new MutableLiveData<>();
    private final MutableLiveData<List<Type>> expenseTypes = new MutableLiveData<>();
    private final MutableLiveData<List<Type>> incomeTypes = new MutableLiveData<>();
    private final MutableLiveData<List<Account>> accounts = new MutableLiveData<>();
    private boolean isNewBill;


    public BillEditViewModel(AppRepository repository) {
        mRepository = repository;
        mRepository.getTypes(true, list -> expenseTypes.setValue(list));
        mRepository.getTypes(false, list -> incomeTypes.setValue(list));
        mRepository.getAccounts(accountList -> accounts.setValue(accountList));
    }

    public void setBill(@NonNull Bill bill) {
        if (bill.getDate() == null) {
            isNewBill = true;
            bill.setDate(DateTime.now());
        } else {
            isNewBill = false;
        }
        mRepository.getAccount(bill.getAccountId(), this::setAccount);
        mRepository.getType(bill.getTypeId(), this::setType);
        balance.setValue(bill.getBalance() == null ? null : bill.getBalance().toString());
        date.setValue(bill.getDate());
        remark.setValue(bill.getRemark());
        billId.setValue(bill.getUuid());
    }

    /**
     * @param account 设置的账户
     */
    public void setAccount(@NonNull Account account) {
        accountImg.setValue(account.getBitmap());
        accountColor.setValue(account.getColorId());
        accountId.setValue(account.getUuid());
    }

    /**
     * @param type 设置类型
     */
    public void setType(@NonNull Type type) {
        isExpense.setValue(type.getIsExpense());
        typeName.setValue(type.getName());
        typeImg.setValue(Type.FOLDER + type.getAssetsName());
        typeId.setValue(type.getUUID());
    }


    public boolean isNewBill() {
        return isNewBill;
    }

    /**
     * @return 类型名
     */
    public MutableLiveData<String> getTypeName() {
        return typeName;
    }

    /**
     * @return 类型图id
     */
    public MutableLiveData<String> getTypeImg() {
        return typeImg;
    }

    /**
     * @return 账户图id
     */
    public MutableLiveData<String> getAccountImg() {
        return accountImg;
    }

    /**
     * @return 账户背景色值
     */
    public MutableLiveData<Integer> getAccountColor() {
        return accountColor;
    }

    /**
     * @return 账单收支
     */
    public MutableLiveData<String> getBalance() {
        return balance;
    }

    /**
     * @return 账单日期
     */
    public MutableLiveData<DateTime> getDate() {
        return date;
    }

    /**
     * @param date 设置账单日期
     */
    public void setDate(DateTime date) {
        this.date.setValue(date);
    }

    /**
     * @return 账单备注
     */
    public MutableLiveData<String> getRemark() {
        return remark;
    }

    /**
     * 保存账单
     */
    public boolean saveBill() {
        if (TextUtils.isEmpty(balance.getValue())) {
            return false;
        }
        Bill bill = new Bill();
        try {
            BigDecimal r = new BigDecimal(balance.getValue());
            bill.setBalance(r);
            bill.setName(getTypeName().getValue());
            bill.setDate(getDate().getValue());
            bill.setRemark(getRemark().getValue());
            bill.setTypeId(typeId.getValue());
            bill.setAccountId(accountId.getValue());
        } catch (Exception e) {
            //Toast.makeText(mContext, "表达式错误", Toast.LENGTH_SHORT).show();
            return false;
        }
        // 刷新账单数据库
        mRepository.getBill(billId.getValue(), b -> {
            if (b == null) {
                mRepository.addBill(bill);
            } else {
                bill.setId(b.getId());
                bill.setUuid(b.getUuid());
                mRepository.updateBill(bill);
            }
        });
        return true;
    }

    /**
     * @return true为支出类型
     */
    public LiveData<Boolean> getExpense() {
        return isExpense;
    }

    @NonNull
    public LiveData<List<Type>> getTypes(boolean b) {
        return b ? expenseTypes : incomeTypes;
    }

    @NonNull
    public LiveData<List<Account>> getAccounts() {
        return accounts;
    }
}

File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/view_model/BillEditViewModel.kt
package io.github.skywalkerdarren.simpleaccounting.view_model

import android.text.TextUtils
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import io.github.skywalkerdarren.simpleaccounting.model.entity.Account
import io.github.skywalkerdarren.simpleaccounting.model.entity.Bill
import io.github.skywalkerdarren.simpleaccounting.model.entity.Type
import io.github.skywalkerdarren.simpleaccounting.model.repository.AppRepository
import org.joda.time.DateTime
import java.math.BigDecimal

/**
 * @author darren
 * @date 2018/4/4
 */
class BillEditViewModel(private val mRepository: AppRepository) : ViewModel() {

    val isExpense = MutableLiveData<Boolean>(true)
    val types = Transformations.switchMap(isExpense) {
        if (it) {
            expenseTypes
        } else {
            incomeTypes
        }
    }
    val balance = MutableLiveData<String?>()
    val bill = MutableLiveData<Bill>()
    var type = MutableLiveData<Type>()
        set(value) {
            field = value
            bill.value?.typeId = type.value?.uuid ?: return
        }
    var account = MutableLiveData<Account>()
        set(value) {
            field = value
            bill.value?.accountId = type.value?.uuid ?: return
        }
    private val expenseTypes = MutableLiveData<List<Type>>()
    private val incomeTypes = MutableLiveData<List<Type>>()
    private val accounts = MutableLiveData<List<Account>>()

    fun setBill(b: Bill?) {
        // empty bill
        if (b == null) {
            mRepository.getAccounts {
                val account = it?.get(0) ?: return@getAccounts
                this@BillEditViewModel.account.value = account
                mRepository.getTypes(true) { types ->
                    val type = types?.get(0) ?: return@getTypes
                        this@BillEditViewModel.type.value = type
                        bill.value = Bill(type.uuid, account.uuid, DateTime(), type.name)
                }
            }
        } else {
            bill.value = b
            mRepository.getAccount(b.accountId) { this@BillEditViewModel.account.value = it }
            mRepository.getType(b.typeId) { type ->
                this@BillEditViewModel.type.value = type ?: return@getType
            }
            balance.value = if (b.balance == null) null else b.balance.toString()
        }
    }

    /**
     * 保存账单
     */
    fun saveBill(failed: SaveFailed): Boolean {
        val finalBill = bill.value
        val finalAccount = account.value
        val finalType = type.value

        if (TextUtils.isEmpty(balance.value) ||
                finalBill == null ||
                finalAccount == null ||
                finalType == null) {
            failed.saveFailed("账单不能为空")
            return false
        }

        finalBill.accountId = finalAccount.uuid
        finalBill.typeId = finalType.uuid
        finalBill.name = finalType.name
        return try {
            val r = BigDecimal(balance.value)
            if (r == BigDecimal.ZERO) {
                failed.saveFailed("账单不能为0")
                return false
            }
            finalBill.balance = r

            // 刷新账单数据库
            mRepository.getBill(finalBill.uuid) {
                if (it == null) {
                    mRepository.addBill(finalBill)
                } else {
                    finalBill.id = it.id
                    finalBill.uuid = it.uuid
                    mRepository.updateBill(finalBill)
                }
            }
            true
        } catch (e: Exception) {
            failed.saveFailed("表达式错误")
            false
        }
    }

    fun getTypes(b: Boolean): LiveData<List<Type>> {
        return if (b) expenseTypes else incomeTypes
    }

    fun getAccounts(): LiveData<List<Account>> {
        return accounts
    }

    interface SaveFailed {
        fun saveFailed(msg: String?)
    }

    init {
        mRepository.getTypes(true) {
            expenseTypes.value = it
        }
        mRepository.getTypes(false) {
            incomeTypes.value = it
        }
        mRepository.getAccounts {
            this@BillEditViewModel.accounts.value = it
        }
    }
}